### ISSUE TITLE
About batch normalization, BNN implementation and latency calculation in CimLoop

### ISSUE THREAD

**Issue by Adrija-debug** [mit-emze/cimloop#13]
Hi,

Could you please provide some insights on the below points:

1. Is there any way of defining batch normalization in CiMLoop? I am trying to post-process the partial sums of a cim array by shifting and scaling them with some factors. How do I implement this with batch normalization instead of using any additional circuits?

2. Inside the workload folder, we have different DNNs and for each DNN, the histograms for input, weights and outputs for different layers are defined. How do I define the same for a Binary Neural Network (BNN) (eg. ResNet-18 BNN)? 

3. How is latency calculated in CiMLoop? Are there any features to implement parallelism in CiMLoop? Say if two branches are working in parallel, then the delay of the faster branch should be overshadowed by the delay of the slower branch, does CiMLoop account for that? How does it calculate the critical path for the maximum delay?

Regards,
Adrija.

**Comment by tanner-andrulis** [comment]
1. Can you write batchnorm as an Einsum? If so, then you can write a workload spec for it.
2. I would only use two bins; the zero bin (centermost) for zero, and the highest bin for one.
3. CiMLoop calculates throughput rather than latency. Latency for a given layer can be calculated with (# of MACs / throughput). This calculation only assumes steady-state latency, so it will miss initial fill latency. For multiple branches (which I assume you mean multiple DNN layers whose outputs feed to the same later layer on), what you should do is calculate the latency of both layers, then take a max. CiMLoop outputs will generally give you per-layer stats, so you can aggregate these per-layer stats to do the calculation.

Hope these were helpful. Let me know if you have any followon questions!

**Comment by Adrija-debug** [comment]
Thanks for the answers, I have a few follow-on questions.
1. Actually I’m trying to design an adder tree that adds all the inputs (total 64 input bits) and sends that sum to multiple crossbar arrays (each array is 64 by 64, total 8 such arrays) for processing. So, ideally the adder tree’s energy should remain almost constant since it is getting shared by many crossbar arrays. However, in my case its energy is getting scaled by the number of crossbars. Below is the yaml code for that. I have tried different approaches but still its energy does not remain constant. Could you please have a look and let me know where I am going wrong?

![Image](https://github.com/user-attachments/assets/be5116dd-134b-4ecc-b72a-7447ae36faee)

2. In my earlier comment, in the 3rd point, when I mentioned about latency, I was actually trying to correlate the above scenario. Ideally, when the adder tree is calculating the input sum, the crossbar array should be calculating the partial sum in parallel. So if the adder tree takes less time than the crossbar array, then its delay should not come into picture. Does CiMLoop take the greater delay out of these two?
3. I think the function ‘run_mapper’ returns latency which is calculated as cycles * cycle_seconds, where cycle_seconds is equal to GLOBAL_CYCLE_SECONDS while the number of 'cycles' is calculated by CiMLoop. I have seen that the value of the variable ‘cycles’ remains 1 even if we lower the BASE_LATENCY. As per my understanding, if we lower the BASE_LATENCY, it should take more number of cycles to perform the computation with same number of inputs and weights. But the ‘cycles’ variable always returns 1 even after decreasing the BASE_LATENCY. Could you explain why that is happening? When will the number of cycles change?

**Comment by tanner-andrulis** [comment]
1. A `*spatial_must_reuse_inputs` on your array would fix that. Right now it's assuming that each input is sent to each array independently, so the adder tree would be activated 8x, one to send an input to each array.
2. Yep! CiMLoop assumes everything is pipelined and takes the long-pole latency, which will not be the adder tree.
3. The number of cycles shouldn't be tied to base latency. If you lower the base latency, then CiMLoop will assume that the entire system will be run at a higher clock speed, and it will take the same number of cycles. It will take lower overall time because each cycle is shorter.
